### PR TITLE
Remove check of nl-char at the end of buffer and incorrect treatment …

### PR DESCRIPTION
**Pull request by wapmorgan**
…of prompt
Related #9

### CODE REVIEW

**Comment by wapmorgan** [comment]
If it's still has value, there's was no trailing newline-chars in all responses. That's raw log with var_dump of buffer:
```
php > $telnet = Graze\TelnetClient\TelnetClient::factory(); $telnet->connect('127.0.0.1:23');var_dump($telnet->execute('wapmorgan', 'Password: '));
string(6) "buffer"
string(94) "elementary OS Freya
wapmorgan-300V3A-300V4A-300V5A-200A4B-200A5B login: wapmorgan
Password: "
object(Graze\TelnetClient\TelnetResponse)#7 (3) {
  ["isError":protected]=>
  bool(false)
  ["responseText":protected]=>
  string(83) "elementary OS Freya
"apmorgan-300V3A-300V4A-300V5A-200A4B-200A5B login: wapmorgan
  ["promptMatches":protected]=>
  array(1) {
    [0]=>
    string(10) "Password: "
  }
}
php > var_dump($telnet->execute('***here is my password***', '\$'));                                                                                                                   string(6) "buffer"
string(603) "
Last login: Thu Jun 15 18:58:43 MSK 2017 from localhost on pts/2
... 
wapmorgan@wapmorgan-300V3A-300V4A-300V5A-200A4B-200A5B:~$"
object(Graze\TelnetClient\TelnetResponse)#7 (3) {
  ["isError":protected]=>
  bool(false)
  ["responseText":protected]=>
  string(602) "
Last login: Thu Jun 15 18:58:43 MSK 2017 from localhost on pts/2
...
wapmorgan@wapmorgan-300V3A-300V4A-300V5A-200A4B-200A5B:~"
  ["promptMatches":protected]=>
  array(1) {
    [0]=>
    string(1) "$"
  }
}
php > var_dump($telnet->execute('ls -l -1 --color=never', '\$'));                                                                                                        string(6) "buffer"
string(810) " total 48
drwxrwxr-x 3 wapmorgan wapmorgan  4096 июня  15 18:02 Documents
drwxrwxr-x 5 wapmorgan wapmorgan  4096 июня  14 02:14 package-build
drwxrwxr-x 3 wapmorgan wapmorgan  4096 июля  11  2015 Steam
drwxr-xr-x 2 wapmorgan wapmorgan  4096 мая   24  2015 Видео
drwxr-xr-x 2 wapmorgan wapmorgan  4096 июня  14 02:26 Документы
drwxr-xr-x 2 wapmorgan wapmorgan 12288 июня  15 17:54 Загрузки
drwxr-xr-x 2 wapmorgan wapmorgan  4096 июня  14 02:26 Изображения
drwxr-xr-x 2 wapmorgan wapmorgan  4096 июня  14 02:26 Музыка
drwxr-xr-x 2 wapmorgan wapmorgan  4096 мая   24  2015 Общедоступные
drwxr-xr-x 2 wapmorgan wapmorgan  4096 мая   24  2015 Шаблоны
wapmorgan@wapmorgan-300V3A-300V4A-300V5A-200A4B-200A5B:~$"
object(Graze\TelnetClient\TelnetResponse)#7 (3) {
  ["isError":protected]=>
  bool(false)
  ["responseText":protected]=>
  string(809) " total 48
drwxrwxr-x 3 wapmorgan wapmorgan  4096 июня  15 18:02 Documents
drwxrwxr-x 5 wapmorgan wapmorgan  4096 июня  14 02:14 package-build
drwxrwxr-x 3 wapmorgan wapmorgan  4096 июля  11  2015 Steam
drwxr-xr-x 2 wapmorgan wapmorgan  4096 мая   24  2015 Видео
drwxr-xr-x 2 wapmorgan wapmorgan  4096 июня  14 02:26 Документы
drwxr-xr-x 2 wapmorgan wapmorgan 12288 июня  15 17:54 Загрузки
drwxr-xr-x 2 wapmorgan wapmorgan  4096 июня  14 02:26 Изображения
drwxr-xr-x 2 wapmorgan wapmorgan  4096 июня  14 02:26 Музыка
drwxr-xr-x 2 wapmorgan wapmorgan  4096 мая   24  2015 Общедоступные
drwxr-xr-x 2 wapmorgan wapmorgan  4096 мая   24  2015 Шаблоны
wapmorgan@wapmorgan-300V3A-300V4A-300V5A-200A4B-200A5B:~"
  ["promptMatches":protected]=>
  array(1) {
    [0]=>
    string(1) "$"
  }
}

```

File: src/PromptMatcher.php [hunk 1]
```diff
@@ -35,20 +35,15 @@ class PromptMatcher
...
         $matches = [];
         $callback = function ($matchesCallback) use (&$matches) {
             $matches = $matchesCallback;
             // replace matches with an empty string (remove prompt from $subject)
             return '';
         };
-        $pattern = sprintf('/%s%s$/', $prompt, $lineEnding);
 
-        $responseText = preg_replace_callback($pattern, $callback, $subject);
+        $responseText = preg_replace_callback('/'.$prompt.'/', $callback, $subject);
```
Comment (john-n-smith) [inline src/PromptMatcher.php]: Including the `$` at the end of the regex should make the preg less resource intensive. It shouldn't have any effect on the match as this check will have been performed on each iteration of `$subject`.

**Comment by wapmorgan** [comment]
I've updated matcher. Take a look

**Comment by wapmorgan** [comment]
What's about merging?
If you want to support both behaviors, I'd propose a setting like `disableNLCharCheck()`.

**Comment by john-n-smith** [comment]
Cool, thanks for the update. 

This highlights some bad assumptions with how the client was designed. As such, I'd like to release this as 3.0.0 along with some other changes.

I'll keep this branch in place if you want to point composer at it in the meantime.

**Comment by john-n-smith** [comment]
Sorry, just realised this is your own fork, so you can point composer at that.

**Comment by wapmorgan** [comment]
I'd prefer to not create package of a fork. What's about configuring this behavior?

**Comment by john-n-smith** [comment]
There's no need to create a package, you can tell composer to point to your fork: http://vvv.tobiassjosten.net/php/have-composer-use-development-branches/

Or, if you want to add the functionality to this PR and fix the tests, I can package the release as v2.1.0 as a stopgap. 

It could be implemented as a separate `setLineEndingPrompt` method that sets a `lineEndingPrompt` value.
If `lineEndingPrompt` is not null, pass its value to `promptMatcher->isMatch`. If it is null, pass `lineEnding`, as is the current behaviour.

For your use case, you would pass an empty string to `setLineEndingPrompt`.

**Comment by john-n-smith** [comment]
Hi @wapmorgan, since https://github.com/graze/telnet-client/releases/tag/v2.0.1 you can define your own `PromptMatcher` and inject it into `TelentClient::__construct` allowing you to configure this behaviour.

**Pull request closed without merging**